### PR TITLE
Bugfix InvalidOperationException/IndexOutOfRangeException in HttpListener.EndGetContext 

### LINKED ISSUE #107025: Exception System.InvalidOperationException in HttpListener.EndGetContext
### Description

After updating the application from Framework to .Net 8 we occasionally get the following exceptions:

```
System.IndexOutOfRangeException:
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Net.HttpListener.RegisterForDisconnectNotification(HttpListenerSession session, UInt64 connectionId, DisconnectAsyncResult& disconnectResult)
   at System.Net.HttpListener.HandleAuthentication(HttpListenerSession session, RequestContextBase memoryBlob, Boolean& stoleBlob)
   at System.Net.ListenerAsyncResult.IOCompleted(ListenerAsyncResult asyncResult, UInt32 errorCode, UInt32 numBytes)
   at System.Net.HttpListener.EndGetContext(IAsyncResult asyncResult)

System.InvalidOperationException:
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at System.Net.HttpListener.HandleAuthentication(HttpListenerSession session, RequestContextBase memoryBlob, Boolean& stoleBlob)
   at System.Net.ListenerAsyncResult.IOCompleted(ListenerAsyncResult asyncResult, UInt32 errorCode, UInt32 numBytes)
   at System.Net.HttpListener.EndGetContext(IAsyncResult asyncResult)
```
After further investigation in the System.Net.HttpListener sourecode I found that the Dictionary DisconnectResults is not always used in a thread safe manner. It is missing lock around the access of DisconnectResults 

### Reproduction Steps

It is hard to reproduce. It happens randomly

### Expected behavior

No exception

### Actual behavior

Exceptions and possible internal corruption of HttpListener

### Regression?

_No response_

### Known Workarounds

I have fixed version of HttpListener (see patch). If this fix is stable for 1 week I will submit a pull request.

[HttpListener-fix-Operations-that-change-non-concurre.patch](https://github.com/user-attachments/files/16760978/HttpListener-fix-Operations-that-change-non-concurre.patch)

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by pjannesen**
Fixes #107025

### CODE REVIEW

**Comment by pjannesen** [comment]
> @dotnet-policy-service agree [company="{your company}"]

@dotnet-policy-service agree [company="Jannesen"]

**Comment by pjannesen** [comment]
> @dotnet-policy-service agree company="your company"



> @pjannesen the command you issued was incorrect. Please try again.
> 
> Examples are:
> 
> ```
> @dotnet-policy-service agree
> ```
> 
> and
> 
> ```
> @dotnet-policy-service agree company="your company"
> ```
@dotnet-policy-service agree company="Jannesen"

File: src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs [hunk 1]
```diff
@@ -70,7 +70,9 @@ public bool UnsafeConnectionNtlmAuthentication
                 {
                     return;
                 }
-                lock ((DisconnectResults as ICollection).SyncRoot)
+
+                var disconnectResults = DisconnectResults;
```
Comment (huoyaoyuan) [inline src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs:74]: The local variable isn't helpful as it doesn't contain any common expression.

Comment (pjannesen) [reply src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs:74]: Definition of DisconnectResults 
``` c#
private Dictionary<ulong, DisconnectAsyncResult> DisconnectResults =>
    LazyInitializer.EnsureInitialized(ref _disconnectResults, () => new Dictionary<ulong, DisconnectAsyncResult>());
```

So every access of DisconnectResults is a call to LazyInitializer.EnsureInitialized.
So using the local variable only one call is necessary.

Comment (huoyaoyuan) [reply src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs:74]: HttpListener isn't so optimized. I don't think reducing several volatile reads can result in real world difference. Let's focus on fixing only.

**Comment by liveans** [comment]
Thanks for the contribution! I'll take a look at this PR, asap!

**Review by liveans — APPROVED** [review]
Change LGTM in general, but I also think that if we should consider using `ConcurrentDictionary` instead of `Dictionary` and locking it with `SyncRoot` everywhere. However, that's another PR's concern, IMO.

**Pull request merged**